### ISSUE TITLE
Use CircuitPython socket support

### ISSUE THREAD

**Issue by dhalbert** [adafruit/Adafruit_CircuitPython_asyncio#4]
Change `asyncio.open_connection()` and `asynico.start_server()` to use CircuitPython's socket support.

**Comment by thorangutang** [comment]
Yes please. I believe this would allow a non blocking tcp/udp socket to be opened.

**Comment by furbrain** [comment]
I've tried to fix this with limited success in these functions

- CircuitPython uses SocketPool, which needs to be initialised with a radio. We can import wifi and do `sp = SocketPool(wifi.radio)`, but this limits our ability to use other networking options as and when they appear
- `socketpool.Socket` doesn't accept `setsockopt` function calls, and there are few other cosmetic changes needed to get these functions working well to start with.
- however the socket ultimately gets sent  via `core.IOQueue._enqueue` to `select.poll.register()` (c code) - which then crashes hard. This is probably because the `poll` code expects underlying objects of mp_stream types and the new sockets are very different types. At one point I had to nuke the flash on my pico W to get it functioning again.
- At this point I've given up because it's looking rather hard, and would possibly need code changes elsewhere in the code base
- On the other hand it should be fairly easy to do a solution in pure python - this would involve doing the polling for new data in python, but obviously this would come with a time penalty.

**Comment by furbrain** [comment]
However, the `asyncio.StreamReader` function does seem to at least start to play well with other "file-like" objects, such as UARTs, the usb_cdc.Serial objects and (not tested yet) the `_bleio.CharacteristicBuffer` streams - but see my other issue for more info

**Comment by dhalbert** [comment]
> * CircuitPython uses SocketPool, which needs to be initialised with a radio. We can import wifi and do `sp = SocketPool(wifi.radio)`, but this limits our ability to use other networking options as and when they appear

You could pass in the SocketPool. `adafruit_requests` and other libraries are agnostic about the source of sockets, and take it as a parameter.

**Comment by Commod0re** [comment]
I've taken a few cracks at this and it's really not so simple as that unfortunately. There are a few issues but the biggest thing is that `SocketPool` sockets do not expose a file descriptor, therefore they can't be used with `select` (and therefore also `IOQueue` and therefore also `aiohttp.stream.Stream` etc)

esp-idf's examples around connecting and using non-blocking sockets also depend on having a file descriptor to use with select (see here: https://github.com/espressif/esp-idf/blob/master/examples/protocols/sockets/non_blocking/main/non_blocking_socket_example.c)

other issues, like the strange connection behavior I see with non-blocking sockets on my Qualia S3 with CircuitPython, can be worked around, but this piece is pretty fundamental to this particular problem. SocketPool Sockets' underlying file descriptor need to be exposed via a `fileno()` method so that `select` can work with them before Stream will work with it at all

**Comment by pterricciano** [comment]
I am running on an Metro M4 Aitlift Lite. I was having a problem where trying to connect to a host which was offline would cause everything else to hang and crash the esp firmware (version 1.7.7). The application is running a control loop where keeping the Delta T jitter as low as possible was a goal. I have implemented a refactor of adafruit_miniqtt and everything underneath (connection managers, esp32_spi ...) to work in a asyncio environment. Other tasks will continue to run even while esp is trying connect. More testing is required but it is currently working for my application. Not sure how to make it available because it affects more than MQTT. The crash of the esp firmware is also fixed.

**Comment by akkana** [comment]
Has anyone found a workaround to this yet? I have a project where a Raspberry Pi Pico W (using socketpool) needs to listen on a network socket, but also needs to watch for a button press. It seems like there's no way to poll a socket, and no way to monitor a socket asynchronously, so I'm not sure how I can monitor both the net and the button pin.

**Comment by pterricciano** [comment]
I have a port of the socket class to support async IO for similar reasons. The changes though have percolated up to many other modules, connection manager, polls etc. It was for a metro m4 and an esp32 wifi. I can push  it if any one is interested.
